### PR TITLE
Use generics to improve @types/rosie safety

### PR DESCRIPTION
**Pull request by chrislgrigg**
This uses generics to help ensure your factory objects are compliant with established interfaces.

```typescript
interface MyCoolInterface {
  name: string;
  really: boolean;
  id?: number;
}

// Will error on `'nope'` because it's not boolean
Factory.define<MyCoolInterface>('MyCoolInterface').attr<'really'>('really', 'nope')

// Give give an error because required key `really` is missing
Factory.define<MyCoolInterface>('MyCoolInterface').attrs({ name: 'Cool Thing' });

// myThing will be of the correct type
const myThing = Factory.build<MyCoolInterface>('MyCoolInterface');

interface PersistedCoolInterface extends MyCoolInterface {
  id: number;
}

// Will give an error if the ExtendedCoolInterface does not actually extended MyCoolInterface
Factory.define('PersistedCoolInterface').extend<PersistedCoolInterface, MyCoolInterface>('MyCoolInterface').sequence('id')
```

I still have not personally tested some of the methods, so it could use some more review. The core functionality (class method `define`, `build`, instance method `extend`, `attr`, `attrs`) is working for me. 

I corrected the tests and added an additional case demonstrating `define` with a generic argument. I added some default generics but it also now requires >= TS 2.3, so I'm not sure how to safely merge this without messing up folks relying on the current behavior.

Appreciate any advice or input.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/rosiejs/rosie/blob/master/src/rosie.js

Previously open under #19966. Tests are passing locally now.

### CODE REVIEW

**Comment by subvertallchris** [comment]
Improved the `attr` instance method's signature quite a bit and added some tests. This should have good backwards compatibility. There's still some room to improve some of the methods, like `option`, but I'll handle that when I have the need or someone else can tackle it. If there are no objects, looking forward to seeing it merged.

**Comment by dotbear** [comment]
Hey @subvertallchris - I've recently run into a problem where `MyFactory.build()` seems to require 2 args (attrs and options), even though these are supposed to be optional (Factory.build("MyFactory") correctly sees attr and opts as optional).

I can't easily spot if you fixed this in this PR? If not, then I'll open one for that :)

**Comment by subvertallchris** [comment]
Hey @dotbear! I fixed it. You can see it [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20010/files#diff-d0ba71f8b707f31b2f25239c7cd03d72L190).

**Comment by dotbear** [comment]
@subvertallchris fantastic, thank you! :)

**Comment by subvertallchris** [comment]
@abner, could you approve?

File: types/rosie/index.d.ts [hunk 1]
```diff
@@ -187,9 +210,9 @@ declare namespace rosie {
      * @param {object=} options
      * @return {*}
      */
-    build(attributes: Object, options: Object): Object;
+    build(attributes: { [k in keyof T]?: T[k] }, options?: any): T;
```
Comment (dotbear) [inline types/rosie/index.d.ts]: I can see now that only the keys are optional, not the entire parameter. Would you like to cut a new PR or should I? :)

```ts
build(attributes: { [k in keyof T]?: T[k] }?, options?: any): T;
```

I believe that is the correct syntax

File: types/rosie/index.d.ts [hunk 2]
```diff
@@ -187,9 +210,9 @@ declare namespace rosie {
...
 
-    buildList(size: number, attributes: Object, options: Object): Object[];
+    buildList(size: number, attributes: { [k in keyof T]?: T[k] }, options: any): T[];
```
Comment (dotbear) [inline types/rosie/index.d.ts]: Same goes for here, both `attributes` and `options` are required, whereas they should be optional :)

**Review by dotbear — COMMENTED** [review]
Hey @subvertallchris! Great PR, we are using it in our project now. Unfortunately it doesn't seem to have made `attributes` optional on `MyFactory.build()`. Same goes for `MyFactory.buildList()`.

**Comment by subvertallchris** [comment]
@dotbear Ah, my apologies! Yes, please do that if you don’t mind.

**Pull request merged**